### PR TITLE
Format response value by looking it up in the config or via the API

### LINKED ISSUE #7: Improved formatting for response
As a user I want two things:

1) The option to put in a configuration the response I'm going to expect for a certain field and have the Response object pick it up automatically

2) The possibility of passing a flag with no default and tell it to go and fetch from Pipedrive what the value supplied means

Both these are to fix what the Pipedrive API should be doing behind the scenes. obvs. Nice one again Pipedrive

### PR DESCRIPTION
**Pull request by pitchinvasion**
This is a bit complext but only because Pipedrive stupidly exposes its
    internal IDs again, hopefully doing this will encapsulate all this nonsense
    away and clean up the code on the other side of things.

    Closes #7

### CODE REVIEW

File: lib/pipekit/field_repository.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+      end
+
+      def get_by_name(name)
+        search_fields("name", name)
+      end
+
+      private
+
+      def search_fields(field_element, value)
+        result = request.get.select { |element| element[field_element] == value }
```
Comment (danldb) [inline lib/pipekit/field_repository.rb]: Feels a little bit strange to depend on `request`, an attribute set by the parent class. It's not obvious that to reuse this module we need to ensure a `#request` method is present in the parent class. Is that a common pattern for composition with modules?

File: lib/pipekit/repository.rb [hunk 2]
```diff
@@ -36,7 +39,7 @@ def where(options)
     #
     # Returns a Hash or nil if none found.
     def find_by(options)
-      where(options).first
+      where(options, true).first
```
Comment (danldb) [inline lib/pipekit/repository.rb]: So this is the bit that may be dangerous - we should check that pipedrive is not automatically making  fuzzy searches.

File: lib/pipekit/request.rb [hunk 3]
```diff
@@ -32,48 +32,66 @@ def initialize(resource)
...
+      return result.response(resource) unless result.fetch_next_request?
+      _get(id, query, result + get_request(id, query, result.next_start))
     end
 
-    def result_from(response)
-      data = response["data"]
-      success = response["success"]
+    def get_request(id, query, start = 0)
+      response = self.class.get(uri(id), options(query: {limit: pagination_limit, start: start}.merge(query)))
+      Result.new(response)
```
Comment (danldb) [inline lib/pipekit/request.rb]: That's cool - I like using the Result object to understand when all the data has been gathered.

File: spec/pipekit/request_spec.rb [hunk 4]
```diff
@@ -62,28 +71,65 @@ module Pipekit
...
-        .to_return(status: 200, body: {success: true}.to_json)
+        .to_return(status: 200, body: {success: true, data: "not empty"}.to_json)
     end
 
     def stub_post(uri, body)
-      stub_request(:post, "#{Pipekit::Request::PIPEDRIVE_URL}/#{uri}")
+      stub_request(:post, "#{Request::PIPEDRIVE_URL}/#{uri}?api_token=")
         .with(body: body)
-        .to_return(status: 200, body: {success: true}.to_json)
+        .to_return(status: 200, body: {success: true, data: "not empty"}.to_json)
```
Comment (danldb) [inline spec/pipekit/request_spec.rb]: We could consider using VCR to use actual response data as a fixture. Could make it a bit easier to debug pipedrive weirdness.

**Comment by danldb** [comment]
Love the churning! :shipit: from me - I would like to look at refactoring the field_repository to not depend on a variable assigned in a different module, but it doesn't seem urgent.

File: lib/pipekit/field_repository.rb (hunk 1 above)
Comment (pitchinvasion) [inline lib/pipekit/field_repository.rb]: Urgh good question will have to see what's out there in the wild, although I suspect it does actually happen, I'd be interested to see how something like MongoId does it as they're normally a decent code base

File: lib/pipekit/repository.rb (hunk 2 above)
Comment (pitchinvasion) [inline lib/pipekit/repository.rb]: well the gem does, definitely something worth looking over agreed

File: spec/pipekit/request_spec.rb (hunk 4 above)
Comment (pitchinvasion) [inline spec/pipekit/request_spec.rb]: yeah it's another good point, in the applications repo I've used a JSON dump which has worked really well, I wonder if we could do that instead, as generally I quite like webmock.

**Comment by pitchinvasion** [comment]
Cool thanks, good points made as well, all of them worth considering if we get the time!

**Pull request merged**